src/target/riscv/riscv-013.c

```diff
@@ -4263,9 +4343,31 @@ static int read_memory_progbuf(struct target *target, target_addr_t address,
 
 	if (mstatus != mstatus_old &&
 			register_write_direct(target, GDB_REGNO_MSTATUS, mstatus_old) != ERROR_OK)
-		return ERROR_FAIL;
+		return MEM_ACCESS_FAILED;
 
-	return result;
+	return (result == ERROR_OK) ? MEM_ACCESS_OK : MEM_ACCESS_FAILED;
+}
+
+static mem_access_result_t
+read_memory_sysbus(struct target *target, target_addr_t address,
+	uint32_t size, uint32_t count, uint8_t *buffer, uint32_t increment)
+{
+	mem_access_result_t skip_reason =
+		mem_should_skip_sysbus(target, address, size, increment, /* is_read = */ true);
+	if (skip_reason != MEM_ACCESS_OK)
+		return skip_reason;
+
+	int ret = ERROR_FAIL;
+	uint64_t sbver = get_field(get_info(target)->sbcs, DM_SBCS_SBVERSION);
+	if (sbver == 0)
+		ret = read_memory_bus_v0(target, address, size, count, buffer, increment);
+	else if (sbver == 1)
+		ret = read_memory_bus_v1(target, address, size, count, buffer, increment);
+	else
+		LOG_TARGET_ERROR(target,
+			"Unknown system bus version: %" PRIu64, sbver);
```

Comment:
Detail: Could you try whether `checkpatch` would permit braces in this if-else cascade? 

If so, it'd improve readability.

```suggestion
	if (sbver == 0) {
		ret = read_memory_bus_v0(target, address, size, count, buffer, increment);
	} else if (sbver == 1) {
		ret = read_memory_bus_v1(target, address, size, count, buffer, increment);
	} else  {
		LOG_TARGET_ERROR(target,
			"Unknown system bus version: %" PRIu64, sbver);
	}
```

Reply:
Addressed